### PR TITLE
CXP-1245: Proposal on validation refactoring

### PR DESCRIPTION
**Pull request by tseho**
**Description (for Contributor and Core Developer)**

Basically, I moved from
```
            new Assert\Sequentially([
                new AttributeTextCriterionStructure(),
                new AttributeTextCriterionValues(),
            ]),
```
to 
```
            new Assert\Sequentially([
                new Assert\Collection([
                    'fields' => [
                        'field' => [
                            new Assert\Type('string'),
                            new Assert\NotBlank(),
                        ],
                        'operator' => [
                            new Assert\Type('string'),
                            new Assert\NotBlank(),
                            new Assert\Choice(['=', '!=', 'CONTAINS', 'DOES NOT CONTAIN', 'STARTS WITH', 'EMPTY', 'NOT EMPTY']),
                        ],
                        'value' => [
                            new Assert\Type('string'),
                        ],
                        'scope' => [
                            new Assert\Type('string'),
                        ],
                        'locale' => [
                            new Assert\Type('string'),
                        ],
                    ],
                    'allowMissingFields' => false,
                    'allowExtraFields' => false,
                ]),
                new CriterionOperatorsRequireEmptyValue([
                    'EMPTY',
                    'NOT_EMPTY',
                ]),
                new AttributeCriterionContainsValidScope(),
                new AttributeCriterionContainsValidLocale(),
            ]),
```

Where everything from the previous `AttributeTextCriterionValues` is now reusable.

Names of the new constraints are not final, I'm open to better suggestions :)

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Comment by tseho** [comment]
`CriterionOperatorsWithEmptyValue` => `CriterionOperatorsRequireEmptyValue` ?
`AttributeCriterionScope` => `AttributeCriterionContainsValidScope` ?
`AttributeCriterionLocale` => `AttributeCriterionContainsValidLocale` ?

File: components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+
+    private function isEmpty(mixed $value): bool
+    {
+        if (null === $value) {
+            return true;
+        }
+
+        if (is_string($value) && '' === $value) {
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php:59]: In `NotBlankValidator`, Symfony uses this condition:
```
if (false === $value || (empty($value) && '0' != $value))
```
What do you think about it?

File: components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        /** @var AttributeCriterion $value */
+
+        if (!$constraint instanceof CriterionOperatorsWithEmptyValue) {
+            throw new UnexpectedTypeException($constraint, CriterionOperatorsWithEmptyValue::class);
+        }
+
+        if (0 === \count($constraint->operators)) {
+            throw new \LogicException('The operators option should not be empty');
+        }
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php:36]: What about validation if there is only operators that need non empty value?

File: components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php (hunk 1 above)
Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php:59]: Yes, I've looked into `NotBlankValidator`. I don't want to use `empty` and for boolean, I'd rather wait for the boolean attribute to check what's the expected behavior.

File: components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php (hunk 2 above)
Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php:36]: In this case, you don't need this validator.
If all operators require NOT NULL, you can enforce it on the `Structure` constraint, with a `NotBlank` on values.
If all operators require NULL, you can also enforice it on the `Structure` constraint.

Comment (masacc) [reply components/catalogs/back/src/Infrastructure/Validation/ProductSelection/CriterionOperatorsWithEmptyValueValidator.php:36]: yes good catch

**Review by masacc — APPROVED** [review]

**Review by guillaumefer-akeneo — APPROVED** [review]

**Pull request merged**